File: src/namelex/phonetics.ts
/** Cologne phonetics (Kölner Phonetik, Postel 1969) for German names. */

import { normalize } from "./normalize.js";

const VOWELS = new Set("aeijouy");
const DT_SIBILANTS = new Set("csz");
const X_PRECEDERS = new Set("ckq");
const C_SIBILANTS = new Set("sz");
const C_FOLLOWERS_INITIAL = new Set("ahkloqrux");
const C_FOLLOWERS_INNER = new Set("ahkoqux");

export function cologne(name: string): string {
  const text = normalize(name);
  if (!text) return "";

  const codes: string[] = [];
  const length = text.length;
  for (let i = 0; i < length; i++) {
    const char = text[i]!;
    const nxt = i + 1 < length ? text[i + 1]! : "";
    const prev = i > 0 ? text[i - 1]! : "";
    let code: string | null = null;

    if (VOWELS.has(char)) code = "0";
    else if (char === "h") continue;
    else if (char === "b") code = "1";
    else if (char === "p") code = nxt === "h" ? "3" : "1";
    else if (char === "d" || char === "t") code = DT_SIBILANTS.has(nxt) ? "8" : "2";
    else if ("fvw".includes(char)) code = "3";
    else if ("gkq".includes(char)) code = "4";
    else if (char === "c") {
      if (i === 0) code = C_FOLLOWERS_INITIAL.has(nxt) ? "4" : "8";
      else if (C_SIBILANTS.has(prev)) code = "8";
      else code = C_FOLLOWERS_INNER.has(nxt) ? "4" : "8";
    } else if (char === "x") code = X_PRECEDERS.has(prev) ? "8" : "48";
    else if (char === "l") code = "5";
    else if (char === "m" || char === "n") code = "6";
    else if (char === "r") code = "7";
    else if (char === "s" || char === "z") code = "8";
    else continue;

    codes.push(code);
  }

  const flat = codes.join("");
  const deduped: string[] = [];
  for (const digit of flat) {
    if (deduped.length === 0 || deduped[deduped.length - 1] !== digit) {
      deduped.push(digit);
    }
  }
  if (deduped.length === 0) return "";
  const [head, ...rest] = deduped;
  return head! + rest.filter((d) => d !== "0").join("");
}

File: packages/namelex/src/namelex/phonetics.py
"""Phonetic keys for German names.

Cologne phonetics (Koelner Phonetik, Postel 1969) is used instead of Soundex:
it was designed for German orthography, handles digraphs such as "sch" and
"ch" and produces variable-length codes.
"""
from __future__ import annotations

from .normalize import normalize

_VOWELS = set("aeijouy")
_DT_SIBILANTS = set("csz")
_X_PRECEDERS = set("ckq")
_C_SIBILANTS = set("sz")
_C_FOLLOWERS_INITIAL = set("ahkloqrux")
_C_FOLLOWERS_INNER = set("ahkoqux")


def cologne(name: str) -> str:
    """Return the Cologne phonetic code of a name.

    >>> cologne("Mueller-Luedenscheidt")
    '65752682'
    >>> cologne("Breschnew")
    '17863'
    """
    text = normalize(name)
    if not text:
        return ""

    codes: list[str] = []
    length = len(text)
    for i, char in enumerate(text):
        nxt = text[i + 1] if i + 1 < length else ""
        prev = text[i - 1] if i > 0 else ""

        if char in _VOWELS:
            code = "0"
        elif char == "h":
            continue
        elif char == "b":
            code = "1"
        elif char == "p":
            code = "3" if nxt == "h" else "1"
        elif char in "dt":
            code = "8" if nxt in _DT_SIBILANTS else "2"
        elif char in "fvw":
            code = "3"
        elif char in "gkq":
            code = "4"
        elif char == "c":
            if i == 0:
                code = "4" if nxt in _C_FOLLOWERS_INITIAL else "8"
            elif prev in _C_SIBILANTS:
                code = "8"
            else:
                code = "4" if nxt in _C_FOLLOWERS_INNER else "8"
        elif char == "x":
            code = "8" if prev in _X_PRECEDERS else "48"
        elif char == "l":
            code = "5"
        elif char in "mn":
            code = "6"
        elif char == "r":
            code = "7"
        elif char in "sz":
            code = "8"
        else:
            continue
        codes.append(code)

    flat = "".join(codes)

    # Collapse repeated digits, then drop every "0" except a leading one.
    deduped: list[str] = []
    for digit in flat:
        if not deduped or deduped[-1] != digit:
            deduped.append(digit)
    if not deduped:
        return ""
    head, tail = deduped[0], [d for d in deduped[1:] if d != "0"]
    return head + "".join(tail)


def cologne_core(name: str) -> str:
    """Cologne code of the name without its nobiliary particle."""
    return cologne(name)


def phonetic_keys(name: str) -> dict[str, str]:
    """All blocking keys used by the lexicon."""
    from .normalize import skeleton

    return {
        "norm": normalize(name),
        "skeleton": skeleton(name),
        "cologne": cologne(name),
    }
